Vimwiki checkboxes

kaymmm opened the pull request:
This addresses #61. It is enabled by default, but perhaps it would be better to disable it by default since it is not standard GFM markdown syntax.

Note: this also fixes a bug in parsing checkboxes and renumbering lists containing checkbox bullets

kaymmm commented:
Hey @dkarter, I'm sure you're busy, but I just wanted to flag this PR again in case it got lost. In addition to the feature additions, it also fixes some bugs I found in the existing codebase.

dkarter commented on spec/renumber_bullets_spec.rb:
```diff
@@ -29,6 +29,44 @@
     TEXT
   end
 
+  it 'renumbers a list containing checkboxes' do
+    filename = "#{SecureRandom.hex(6)}.txt"
+    write_file(filename, <<-TEXT)
+      # Hello there
+      33. this is the first bullet
+      - [x] this is the second bullet
+      1.     this is the third bullet
+        - [ ] this is the fourth bullet
+      4. this is the fifth bullet
+
+      - [o] second bullet list
+      a. second item
+      - [x] third item
+    TEXT
+
+    vim.edit filename
+    vim.type 'j'
+    vim.feedkeys 'gN'
+    vim.type '}j'
+    vim.feedkeys 'gN'
+    vim.write
+
+    file_contents = IO.read(filename)
+
+    expect(file_contents).to eq normalize_string_indent(<<-TEXT)
+      # Hello there
+      1.  this is the first bullet
+      2.  this is the second bullet
+      3.  this is the third bullet
+        - [ ] this is the fourth bullet
+      4.  this is the fifth bullet
+
+      - [o] second bullet list
+      - [ ] second item
+      - [x] third item\n
+    TEXT
+  end
```
There are a few things that don't make too much sense to me about this one:
1. Why do the renumbered bullets have 2 spaces between the `.` and their content (I think it should be a single space, just like the checkboxes)?
2. Why did `- [x] this is the second bullet` turn into a numeric bullet (`2.  this is the second bullet`)? It seems like you were following the indentation level + parent bullet type to determine whether to "renumber" or not, but I don't think we should change the author's intention of bullet type, just a literal renumber.

I expected the result to look more like:

```
1. this is the first bullet
- [x] this is the second bullet
2.     this is the third bullet
  - [ ] this is the fourth bullet
3. this is the fifth bullet
    
- [o] second bullet list
a. second item
- [x] third item
```

While your expected result looks more "formatted" and pretty, it changes the meaning of the list items and not as true to it's source and the idea of "renumbering".

kaymmm replied:
>Why do the renumbered bullets have 2 spaces between the . and their content (I think it should be a single space, just like the checkboxes)?

This is because the original list was numbered with a `33.` so it defines the bullet length as (3+3+.+' ')=4, thus renumbered bullets maintain this bullet length. I agree that it's a bit weird, and perhaps renumbering should use the spaces after bullet instead of bullet length.

> Why did - [x] this is the second bullet turn into a numeric bullet (2. this is the second bullet)? It seems like you were following the indentation level + parent bullet type to determine whether to "renumber" or not, but I don't think we should change the author's intention of bullet type, just a literal renumber.

I went back and forth on this one for exactly the reasons you stated. I ultimately decided to renumber based on indentation level alone, and not bullet type because when one changes the indentation level and it automatically renumbers, the expected behavior is for the bullet type to change to match the other bullets at the same indentation level. It would be an easy fix to change this behavior. Perhaps a better way of handling this would be to just treat checkboxes differently from other bullet types and ignore them completely in renumbering.

dkarter replied:
> and perhaps renumbering should use the spaces after bullet instead of bullet length.

Agreed. We should only change the number/letter when using the command, not the content of the bullet. 

> It would be an easy fix to change this behavior. Perhaps a better way of handling this would be to just treat checkboxes differently from other bullet types and ignore them completely in renumbering.

That would be ideal since checkbox don’t represent any numerical ordering by nature. To me it makes sense to keep them unchanged by ‘renumber’.

kaymmm replied:
Changes added to the PR! This also fixes a bug that I previously introduced in resolving rom/abc.

kaymmm commented:
@dkarter, I addressed the issues with renumbering and the spaces after the bullet closure. Let me know if you catch any other bugs or odd behavior. I'm happy to make additional changes.

dkarter commented:
Thank you for addressing the issue @kaymmm!

The feature looks awesome! I might just start using NVIM + bullets as my TODO list :)

I'm almost ready to merge this, but I noticed another oddity with regards to trailing space after bullet if special characters are used. This does not seem to happen when regular characters are used:
![2020-04-27 12 11 47](https://user-images.githubusercontent.com/551858/80400462-7480a200-8880-11ea-9323-c01db2073b1b.gif)

kaymmm commented:
> I'm almost ready to merge this, but I noticed another oddity with regards to trailing space after bullet if special characters are used. This does not seem to happen when regular characters are used:

Oof, I think I know exactly what the problem is. The special characters are multi-byte, double width UTF characters so the bullet length is miscalculated, and pad_length adds extra trailing spaces. I'll try to put together a fix for this later today if I have a chance.

kaymmm commented:
This commit should fix the padding bug.

dkarter commented on plugin/bullets.vim:
```diff
@@ -421,7 +421,8 @@ fun! s:next_num_bullet(bullet)
 fun! s:next_chk_bullet(bullet)
-  return '- [ ]'
+  let l:checkbox_markers = split(g:bullets_checkbox_markers, '\zs')
+  return '- [' . l:checkbox_markers[0] . ']'
```
💯

dkarter commented:
Fantastic work mate! This is amazing, thank you so much for addressing my feedback and for your continued contribution! 💙💚💛
